### PR TITLE
fix: currentItem showing error in debugger for list widgets

### LINKED ISSUE #8677: [Bug] currentItem showing error in List widget

### PR DESCRIPTION
**Pull request by ohansFavour**
This PR introduces the concept of private widgets. To prevent wasted evaluations, these widgets do not get evaluated.
For example, for widget `Button1` in a List widget `List1`, `List1.template.Button1.text` gets evaluated, so there is no need to evaluate `Button1.text`. 

## Description

Fixes #8677

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

- Manual

- Jest

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: fix/currentItem-showing-error-in-debugger-for-list-widget 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.06 **(0.05)** | 36.79 **(0.19)** | 34.88 **(0.06)** | 55.55 **(0.05)**
 :green_circle: | app/client/src/sagas/WidgetOperationSagas.tsx | 51.54 **(0.22)** | 39.02 **(0.48)** | 52.08 **(0)** | 52.79 **(0)**
 :green_circle: | app/client/src/utils/DSLMigrations.ts | 73.18 **(0.49)** | 69.62 **(6.18)** | 71.13 **(0.92)** | 72.99 **(0.52)**
 :green_circle: | app/client/src/utils/DynamicBindingUtils.ts | 84.07 **(0.55)** | 71.43 **(1.79)** | 75.86 **(0)** | 83.44 **(0.62)**
 :green_circle: | app/client/src/widgets/BaseWidget.tsx | 78.86 **(4.88)** | 44.44 **(8.88)** | 75 **(6.25)** | 77.97 **(5.09)**
 :green_circle: | app/client/src/widgets/ListWidget/widget/index.tsx | 63.43 **(1.48)** | 27.14 **(1.25)** | 58.49 **(1.63)** | 62.41 **(1.62)**
 :green_circle: | app/client/src/workers/DataTreeEvaluator.ts | 77.9 **(0.07)** | 61.27 **(0.12)** | 76.4 **(0.54)** | 77.66 **(0.07)**
 :green_circle: | app/client/src/workers/evaluationUtils.ts | 58.15 **(0.69)** | 54.49 **(0.55)** | 62.96 **(0.7)** | 55.91 **(0.72)**</details>

### CODE REVIEW

File: app/client/src/utils/DSLMigrations.ts [hunk 1]
```diff
@@ -83,6 +83,39 @@ const addLogBlackListToAllListWidgetChildren = (
...
+ * @returns
+ */
+const addPrivateWidgetsToAllListWidgets = (
+  currentDSL: ContainerWidgetProps<WidgetProps>,
+) => {
+  currentDSL.children = currentDSL.children?.map((children: WidgetProps) => {
+    if (children.type === "LIST_WIDGET") {
+      const widgets = get(
+        children,
+        "children.0.children.0.children.0.children",
```
Comment (Rishabh-Rathod) [inline app/client/src/utils/DSLMigrations.ts:99]: Please add a comment for understanding why `children.0.children.0.children.0.children` is accessed.

Comment (ohansFavour) [reply app/client/src/utils/DSLMigrations.ts:102]: > Please add a comment for understanding why `children.0.children.0.children.0.children` is accessed.

Done

File: app/client/src/workers/DataTreeEvaluator.ts [hunk 2]
```diff
@@ -79,6 +80,7 @@ export default class DataTreeEvaluator {
   widgetConfigMap: WidgetTypeConfigMap = {};
   evalTree: DataTree = {};
   allKeys: Record<string, true> = {};
+  privateWidgets: Record<string, true> = {};
```
Comment (Rishabh-Rathod) [inline app/client/src/workers/DataTreeEvaluator.ts:83]: Can we create a type for privateWidget and add a comment there on what it means?
```js
/** 
these widgets do not get evaluated
*/
type PrivateWidgets = Record<string, true>;```

**Review by Rishabh-Rathod — APPROVED** [review]
Looks good. 
Few nitpicks.

**Review by Rishabh-Rathod — COMMENTED** [review]
Also, can we add a test for privateWidgets specifically?  
1. `addPrivateWidgetsToAllListWidgets`
2. `isPrivateEntityPath`

**Comment by ohansFavour** [comment]
> Also, can we add a test for privateWidgets specifically?
> 
> 1. `addPrivateWidgetsToAllListWidgets`
> 2. `isPrivateEntityPath`

Done

**Review by Rishabh-Rathod — CHANGES_REQUESTED** [review]
Need to re-review, not sure why the older review wasn't considered stale after the commit.

**Review by Rishabh-Rathod — APPROVED** [review]
Looks good to me

**Pull request merged**